tests/std/tests/P2505R5_monadic_functions_for_std_expected/test.cpp

```diff
@@ -0,0 +1,377 @@
+    const auto immov   = [](auto...) { return Immovable{88}; };
+    const auto to_void = [](auto...) { return; };
+
+    {
+        decltype(auto) result = forward<Expected>(engaged).transform(f);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(result == 55);
+    }
+    {
+        decltype(auto) result = forward<Expected>(unengaged).transform(f);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(!result);
+        assert(result.error() == 22);
+    }
+
+    if constexpr (!is_void_v<Val>) {
+        {
+            decltype(auto) result = forward<Expected>(engaged).transform(&Thingy::member_func);
+            static_assert(is_same_v<decltype(result), expected<int, int>>);
+            assert(result == 66);
+        }
+        {
+            decltype(auto) result = forward<Expected>(unengaged).transform(&Thingy::member_func);
+            static_assert(is_same_v<decltype(result), expected<int, int>>);
+            assert(result.error() == 22);
+        }
+    }
+
+    if constexpr (!is_permissive) { // TRANSITION, VSO-1734935
+        {
+            decltype(auto) result = forward<Expected>(engaged).transform(immov);
+            static_assert(is_same_v<decltype(result), expected<Immovable, int>>);
+            assert(result->v == 88);
+        }
+        {
+            decltype(auto) result = forward<Expected>(unengaged).transform(immov);
+            static_assert(is_same_v<decltype(result), expected<Immovable, int>>);
+            assert(!result);
+            assert(result.error() == 22);
+        }
+    }
+
+    {
+        decltype(auto) result = forward<Expected>(engaged).transform(to_void);
+        static_assert(is_same_v<decltype(result), expected<void, int>>);
+        assert(result);
+    }
+    {
+        decltype(auto) result = forward<Expected>(unengaged).transform(to_void);
+        static_assert(is_same_v<decltype(result), expected<void, int>>);
+        assert(!result);
+        assert(result.error() == 22);
+    }
+
+    const auto to_thingy = [](int i) { return Thingy{i}; };
+
+    {
+        decltype(auto) result = forward<Expected>(engaged).transform_error(to_thingy);
+        static_assert(is_same_v<decltype(result), expected<Val, Thingy>>);
+        assert(result);
+        if constexpr (!is_void_v<Val>) {
+            assert(result->x == 11);
+        }
+    }
+    {
+        decltype(auto) result = forward<Expected>(unengaged).transform_error(to_thingy);
+        static_assert(is_same_v<decltype(result), expected<Val, Thingy>>);
+        assert(!result);
+        assert(result.error().x == 22);
+    }
+    {
+        decltype(auto) result =
+            forward<Expected>(engaged).transform_error(to_thingy).transform_error(&Thingy::member_func);
+        static_assert(is_same_v<decltype(result), expected<Val, int>>);
+        if constexpr (!is_void_v<Val>) {
+            assert(result->x == 11);
+        }

```

Comment:
When `is_void_v<Val>`, this isn't inspecting `result` at runtime at all. It seems that we should `assert(result);`, as line 176 does above. Occurs below.